token/services/sql/tokens.go

```diff
@@ -0,0 +1,886 @@
+	if _, err := db.db.Exec(query, tr.Namespace, tr.TxID, tr.Index, tr.IssuerRaw, tr.OwnerRaw, tr.Ledger, tr.LedgerMetadata, tr.Type, tr.Quantity, tr.Amount, now); err != nil {
+		logger.Errorf("error storing token [%s] in table [%s]: [%s][%s]", tr.TxID, table, err, string(debug.Stack()))
+		return errors.Wrapf(err, "error storing token [%s] in table [%s]", tr.TxID, table)
+	}
+
+	// Store ledger part
+	if ledgerInsert {
+		query = fmt.Sprintf("INSERT INTO %s (ns, tx_id, idx, ledger, ledger_metadata) VALUES ($1, $2, $3, $4, $5)", db.table.Ledger)
+		logger.Debug(query, tr.Namespace, tr.TxID, tr.Index, len(tr.Ledger), len(tr.LedgerMetadata))
+		if _, err := db.db.Exec(query, tr.Namespace, tr.TxID, tr.Index, tr.Ledger, tr.LedgerMetadata); err != nil {
+			// is the token already here?
+			ids := []*token.ID{{TxId: tr.TxID, Index: tr.Index}}
+			logger.Debugf("retrieve ledger tokens for [%s][%s]", tr.Namespace, ids)
+			args := make([]interface{}, 0)
+			where := whereTokenIDs(&args, tr.Namespace, ids)
+			query := fmt.Sprintf("SELECT tx_id, idx FROM %s WHERE %s", db.table.Ledger, where)
+			logger.Debug(query, args)
+			row := db.db.QueryRow(query, args...)
+			var txID string
+			var index int
+			if err2 := row.Scan(&txID, &index); err2 != nil {
+				if errors.Is(err2, sql.ErrNoRows) {
+					logger.Errorf("error storing ledger token [%s] in table [%s]: [%s][%s]", tr.TxID, db.table.Ledger, err2, string(debug.Stack()))
+					return errors.Wrapf(err, "error storing ledger token [%s] in table [%s]", tr.TxID, db.table.Ledger)
+				}
+			}
+			// this was already inserted, skip
+		}
+	}
+
+	// Store ownership
+	for _, eid := range owners {
+		query = fmt.Sprintf("INSERT INTO %s (ns, tx_id, idx, enrollment_id) VALUES ($1, $2, $3, $4)", db.table.Ownership)
+		logger.Debug(query, tr.Namespace, tr.TxID, tr.Index, eid)
+		if _, err := db.db.Exec(query, tr.Namespace, tr.TxID, tr.Index, eid); err != nil {
+			return errors.Wrapf(err, "error storing token ownership [%s]", tr.TxID)
+		}
+	}
+
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, "failed committing token storage")
+	}
+
+	return err
+}
+
+func (db *TokenDB) OwnersOf(ns, txID string, index uint64) (*token.Token, []string, error) {
+	args := make([]interface{}, 0)
+	tokenIDs := []*token.ID{{TxId: txID, Index: index}}
+	where := whereTokenIDs(&args, ns, tokenIDs)
+
+	// select token
+	query := fmt.Sprintf("SELECT owner_raw, token_type, quantity FROM %s WHERE %s AND is_deleted = false", db.table.Tokens, where)
+	logger.Debug(query, args)
+	row := db.db.QueryRow(query, args...)
+	var tokenOwner []byte
+	var tokenType string
+	var quantity string
+	if err := row.Scan(&tokenOwner, &tokenType, &quantity); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, nil, nil
+		}
+		return nil, nil, err
+	}
+
+	// select owners
+	query = fmt.Sprintf("SELECT enrollment_id FROM %s WHERE %s", db.table.Ownership, where)
```

Comment:
token and owners can be retrieved in one query. See UnspentTokensIteratorBy(ns, ownerEID, typ string), it's quite similar (would be good to have a test!).

Reply:
isn't the `join` more expensive?

Reply:
No, the database optimizes for queries like that, especially because we have indexes on the txid on the ownership and token tables. I mean, in this case it's not going to matter that much in terms of performance but I'd prefer to use the db functionalities where we can. Let me know if you want me to write the query.

Reply:
Perfect, I'll change the code 👍

Reply:
@arner , actually, shall we address this when we rework the table organization in another PR?